Upgrade angular10

atarix83 opened the pull request:
## References
Fixes #787

## Description
This is a draft PR to provide the upgrade to angular 10 version.
Here a list of major changes important to know :
- Ivy renderer is now enabled https://next.angular.io/guide/ivy-compatibility#changes-you-may-see
- added Angular compatibility compiler postinstall script (ngcc) to package.json, this make sure node_modules will be compatible with the Ivy renderer. It will be run after a npm install.
- the testing helper function async has been deprecated and renamed waitForAsync
- Lazy loading, because of deprecation, we must now use dynamic imports instead of a string.
```
// Before
loadChildren: './path/lazy.module#LazyModule'
// Now
loadChildren: () => import('./path/lazy.module').then(m => m.LazyModule)
```

## Instructions for Reviewers
Unfortunately there are some unresolved issues and any help on them is well appreciated:
- (HAS WORKAROUND) the dynamic menu works properly only when application is running without SSR (`yarn start:dev`). when application is running with SSR enabled (`yarn start`) it gives several errors: 
![Schermata da 2020-12-04 14-35-26](https://user-images.githubusercontent.com/2486489/101759263-a8d86400-3ad9-11eb-924e-c80cc2d92545.png)
 these are due to the fact that all the dynamic menu components, which use the `rendersSectionForMenu` decorator, are instantiated once only during SSR but not during CSR, so the [menuComponentMap](https://github.com/4Science/dspace-angular/blob/upgrade_angular10/src/app/shared/menu/menu-section.decorator.ts#L3) is empty. Other dynamic components that are using the same logic work properly.
- only during SSR there are the following errors:
```
ERROR Error: inject() must be called from an injection context
    at injectInjectorOnly (/path/node_modules/@angular/core/bundles/core.umd.js:1202:19)
    at Object.ɵɵinject (/path/node_modules/@angular/core/bundles/core.umd.js:1213:61)
    at Object.Angulartics2_Factory [as factory] (/path/node_modules/angulartics2/bundles/angulartics2.umd.js:426:94)
    at R3Injector.hydrate (/path/dist/server/main.js:64990:35)
    at R3Injector.get (/path/dist/server/main.js:64811:33)
    at NgModuleRef$1.get (/path/dist/server/main.js:77951:33)
    at R3Injector.get (/path/dist/server/main.js:64822:33)
    at NgModuleRef$1.get (/path/dist/server/main.js:77951:33)
    at Object.get (/path/dist/server/main.js:77648:35)
    at getOrCreateInjectable (/path/dist/server/main.js:57859:39)

ERROR Error: Multiple components match node with tagname ds-view-tracker
    at throwMultipleComponentError (/path/dist/server/main.js:55881:11)
    at findDirectiveDefMatches (/path/dist/server/main.js:61887:29)
    at resolveDirectives (/path/dist/server/main.js:61679:31)
    at elementStartFirstCreatePass (/path/dist/server/main.js:68439:27)
    at ɵɵelementStart (/path/dist/server/main.js:68477:9)
    at Module.ɵɵelement (/path/dist/server/main.js:68552:5)
    at HomePageComponent_ng_container_4_Template (/path/dist/server/main.js:237754:8)
    at executeTemplate (/path/dist/server/main.js:61147:9)
    at renderView (/path/dist/server/main.js:60954:13)
    at TemplateRef.createEmbeddedView (/path/dist/server/main.js:63752:17)
```
- due to jasmine update it seems that is not possible to create spy on pure function imported from a module or 3rd party package. Where it has been possible I've fixed problematic tests providing the mock store from `@ngrx/store/testing`, but there are several test that need to be rewritten maybe in a followup PR, in my opinion. Now they are temporary disabled. Here the list:
   - src/app/+browse-by/+browse-by-switcher/browse-by-switcher.component.spec.ts
   - src/app/shared/cookies/browser-klaro.service.spec.ts
   - src/app/shared/mydspace-actions/claimed-task/switcher/claimed-task-actions-loader.component.spec.ts
   - src/app/core/cache/builders/link.service.spec.ts
   - src/app/shared/object-collection/shared/listable-object/listable-object-component-loader.component.spec.ts
   - src/app/shared/metadata-representation/metadata-representation-loader.component.spec.ts
   - src/app/core/data/relationship.service.spec.ts

- currently there are several tests that fail only when running test with headless chrome (`yarn run test:headless`) while are successful if running with chrome (`yarn run test`), Here the list : 
  - src/app/+item-page/edit-item-page/item-reinstate/item-reinstate.component.spec.ts
  - src/app/+item-page/edit-item-page/item-withdraw/item-withdraw.component.spec.ts
  - src/app/+item-page/edit-item-page/item-public/item-public.component.spec.ts
  - src/app/+item-page/edit-item-page/item-private/item-private.component.spec.ts
  - src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.component.spec.ts
  - src/app/shared/starts-with/date/starts-with-date.component.spec.ts
- NEW: Circular dependencies introduced.

List of changes in this PR:
* upgraded lib dependencies
* fixed a lot of problematic test
* updated tslint configuration (some rule are outdated) and added a script (`yarn run lint-fix`) to automatically fix the code

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [ ] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

atarix83 commented:
> currently there are several tests that fail only when running test with headless chrome (yarn run test:headless) while are successful if running with chrome (yarn run test), Here the list :
    src/app/+item-page/edit-item-page/item-reinstate/item-reinstate.component.spec.ts
    src/app/+item-page/edit-item-page/item-withdraw/item-withdraw.component.spec.ts
    src/app/+item-page/edit-item-page/item-public/item-public.component.spec.ts
    src/app/+item-page/edit-item-page/item-private/item-private.component.spec.ts
    src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.component.spec.ts
    src/app/shared/starts-with/date/starts-with-date.component.spec.ts

This issue is now fixed

tdonohue reviewed (commented):
@atarix83 : I've started doing a general code review of the changes here.  I'm going to do this review little by little (as it's hard to review nearly 800 files all at once).  In any case, I've finished reviewing the first 200 files & only found one minor issue (see inline below). 

Overall, I see the changes here are very minor in most files (most are only changed to add semicolons on the end of each line).  So, that makes it easier to scan through the changes to just look for files with larger modifications.

atarix83 commented on src/app/+browse-by/+browse-by-switcher/browse-by-switcher.component.spec.ts:
```diff
@@ -4,10 +4,10 @@ import { NO_ERRORS_SCHEMA } from '@angular/core';
 import { environment } from '../../../environments/environment';
 
-describe('BrowseBySwitcherComponent', () => {
+xdescribe('BrowseBySwitcherComponent', () => {
```
yes this is intentional, as I said in the PR description there a list of excluded test

atarix83 commented:
@tdonohue thanks for starting the review

> Overall, I see the changes here are very minor in most files (most are only changed to add semicolons on the end of each line). So, that makes it easier to scan through the changes to just look for files with larger modifications.

Yes since I've updated some tslint rule that are outdated, many changes are nothing more than fixing lint errors

artlowel commented:
The klaro issue can be fixed by applying these changes: 
https://github.com/atmire/dspace-angular/commit/45e5775bfa1f248829c4d59b106203fa234f7cca

It was broken because [klaro v0.7 had breaking changes](https://kiprotect.com/docs/klaro/releases).

atarix83 commented:
> The klaro issue can be fixed by applying these changes:
atmire@45e5775
It was broken because klaro v0.7 had breaking changes.

thanks a lot @artlowel it works now

artlowel commented:
I also looked in to the menu issue, and there the problem seems to be that with the combination of Ivy and AoT [the decorators that specify which component renders which menu section](https://github.com/DSpace/dspace-angular/blob/main/src/app/+admin/admin-sidebar/expandable-admin-sidebar-section/expandable-admin-sidebar-section.component.ts#L23)  and [menu item](https://github.com/DSpace/dspace-angular/blob/main/src/app/shared/menu/menu-item/link-menu-item.component.ts#L15) aren't loaded (in time) on the client. So they can't be found when the menu is rendered.

Perhaps the new way of lazy-loading modules is also involved, because very similar decorators do seem to work for the listableObjects

I spent a while debugging this. I don't have any more time to look in to it further, but if you can't find how to fix it, you can always get rid of the decorators, and instead replace the getComponentForMenu functions with switches, and then create a ticket to fix it later. 

I did so in [this commit](https://github.com/atmire/dspace-angular/commit/a977b798512f337acb607127c52db104e4f8b133), to test the theory that the decorators were at fault, and it seems to work at first sight

atarix83 commented:
thanks @artlowel  I agree with your conclusions.
we have spent enough time on this issue so I won't investigate more on that but I pick your fix up to resolve the problem

tdonohue commented on src/app/shared/testing/SidebarServiceStub.ts:
```diff
@@ -0,0 +1,14 @@
+import {of as observableOf } from 'rxjs';
+
+export class SidebarServiceStub {
```
Tiny thing, could we name this file `sidebar-service.stub.ts`?  That'd keep with the current naming convention, as I see we have other stubs named `*.stub.ts` in this same directory.

atarix83 replied:
done

tdonohue reviewed (commented):
@atarix83 : I've completed a full code review (of all 805 files) in this PR.  As I previously noted, the vast majority of the changes are very minor code syntax changes (correcting semicolons, correctly Array syntax, etc).  **NOTE for other reviewers -- this means that if you don't have time to review all the code, concentrate your effort on reviewing files with >=50 lines changed....others are likely just minor syntax changes)**

I did stumble on a few minor things during the last part of the review.  Please see comments inline below.  Mostly they are questions on commented out specs & other minor things.

atarix83 commented:
@tdonohue @artlowel 

main is now merged into this PR.
I've also resolved issue with circular dependencies and addressed @tdonohue's feedback

atarix83 commented on src/app/shared/menu/menu.service.spec.ts:
```diff
@@ -216,14 +179,14 @@ describe('MenuService', () => {
     describe('when the subsection list is empty', () => {
       beforeEach(() => {
-        selectSpy.and.callFake(() => {
+/*        selectSpy.and.callFake(() => {
```
removed commented part

atarix83 commented on src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/relationship.effects.spec.ts:
```diff
@@ -116,49 +116,50 @@ describe('RelationshipEffects', () => {
+            flush();
+            // TODO check following expectations with the implementation
+            // expect((relationEffects as any).initialActionMap[identifier]).toBe(action.type);
+            // expect((relationEffects as any).debounceMap[identifier].value).toBe(action.type);
```
you're right this should be added to TODO list. Since this class gave me problem after upgrade, I had to fix this it. It is very likely that before those tests had succeeded only because the expectations weren't really checked

tdonohue reviewed (commented):
@atarix83 : I gave this a test today using `yarn start`.  While I didn't test every single feature, I clicked around into many/most and found no major errors/problems.  So, this appears to be working. 

I did see the two SSR `ERROR` messages listed in the description above (so those both still exist), but they seem to not cause issues in the UI behavior itself.  I also did find one additional warning that is logged on command line when running `yarn start`. This seems like it may be possible to quickly resolve as it tells us how to fix the warning.

```
DEPRECATED: DI is instantiating a token "EndpointMapResponseParsingService" that inherits its @Injectable decorator but does not provide one itself.
This will become an error in a future version of Angular. Please add @Injectable() to the "EndpointMapResponseParsingService" class.
```

Overall, in my opinion, this is looking "good enough" to merge soon, provided that @artlowel  agrees.  There are still some outstanding issues & some specs are still disabled (as noted in the description) which means the coverage decreases by about 5%.  But, it's working.  So, if we all agree that the outstanding issues can be moved to separate tickets, then I'm +1 in merging this soon 👍

artlowel commented on src/app/shared/mocks/angulartics.service.mock.ts:
```diff
@@ -1,4 +1,7 @@
+import { Injectable } from '@angular/core';
+
 /* tslint:disable:no-empty */
+@Injectable({providedIn: 'root'})
```
It seems unlikely that a mock service would need providedIn: root

atarix83 replied:
done

artlowel commented on src/main.server.ts:
```diff
@@ -3,6 +3,11 @@ import 'zone.js/dist/zone';
+/******************************************************************
+ * Load `$localize` - used if i18n tags appear in Angular templates.
+ */
+import '@angular/localize/init';
```
As we're using ngx-translate and not Angular's own i18n, I don't think this is necessary

atarix83 replied:
you right but I had to add it anyway otherwise we get the following error during SSR 
```
ERROR Error: It looks like your application or one of its dependencies is using i18n.
Angular 9 introduced a global `$localize()` function that needs to be loaded.
Please run `ng add @angular/localize` from the Angular CLI.
(For non-CLI projects, add `import '@angular/localize/init';` to your `polyfills.ts` file.
For server-side rendering applications add the import to your `main.server.ts` file.)
```

artlowel replied:
I see. It's a [dependency of ng-bootstrap](https://ng-bootstrap.github.io/#/getting-started#i18n) it seems. 

Perhaps change the comment from `used if i18n tags appear in Angular templates`, to: `not used for i18n in this project, we use ngx-translate. It's used for localization of dates, numbers, currencies, etc.` in order to prevent confusion in the future?

artlowel commented on src/polyfills.ts:
```diff
@@ -62,3 +62,8 @@ import 'core-js/features/reflect';
+/******************************************************************
+ * Load `$localize` - used if i18n tags appear in Angular templates.
+ */
+import '@angular/localize/init';
```
As we're using ngx-translate and not Angular's own i18n, I don't think this is necessary

atarix83 replied:
the same as above

artlowel reviewed (commented):
Thanks @atarix83!

I finished my code-review today. I'll do some more testing tomorrow.

atarix83 commented on server.ts:
```diff
@@ -25,135 +26,142 @@ import * as morgan from 'morgan';
+   */
+  server.use('/sitemap**', createProxyMiddleware({ target: `${environment.rest.baseUrl}/sitemaps`, changeOrigin: true }));
+
+  // TODO: implement data requests securely
+  server.get('/api/**', (req, res) => {
+    res.status(404).send('data requests are not yet supported');
```
It's not necessary, removed

atarix83 commented on src/app/+admin/admin-access-control/group-registry/group-form/group-form.component.ts:
```diff
@@ -143,17 +143,18 @@ export class GroupFormComponent implements OnInit, OnDestroy {
     this.canEdit$ = this.groupDataService.getActiveGroup().pipe(
       hasValueOperator(),
       switchMap((group: Group) => {
+        console.log(group);
```
removed

atarix83 commented on tsconfig.server.json:
```diff
@@ -2,8 +2,12 @@
   },
+  "files": [
+    "src/main.server.ts",
+    "server.ts"
```
done

atarix83 commented on src/app/core/services/hard-redirect.service.ts:
```diff
@@ -5,7 +5,7 @@ import { URLCombiner } from '../url-combiner/url-combiner';
  * Service to take care of hard redirects
  */
-@Injectable()
+@Injectable({providedIn: 'root'})
```
done

atarix83 commented on src/app/core/services/server-hard-redirect.service.ts:
```diff
@@ -6,7 +6,7 @@ import { HardRedirectService } from './hard-redirect.service';
  * Service for performing hard redirects within the server app module
  */
-@Injectable()
+@Injectable({providedIn: 'root'})
```
done

atarix83 commented on src/app/core/services/server-cookie.service.ts:
```diff
@@ -1,29 +1,29 @@
 import { CookieService, ICookieService } from './cookie.service';
 
-@Injectable()
+@Injectable({providedIn: 'root'})
```
done

atarix83 commented:
@tdonohue @artlowel 
thanks for the review, I've fixed the deprecated warning and and applied changes to address Art's feedback

artlowel reviewed (changes requested):
I found a few issues while testing:

## Issues
All issues below were discovered in production mode, using the 4science demo server as Rest API

### 1. Browse pages are blank
If you go to any browse page, either on site level or on a collection or community level, the page stays blank

### 2. The layout of the metadata schema and field registries has changed
All fields are shown underneath one another on the metadata schema and field registry pages, on the demo server namespace and name, and element and qualifier are inline


### 3. Person list elements are blank 
If you search for items with Item Type: Person, the results are all black, except for the type label:
http://localhost:4000/search?page=1&query=person&f.entityType=Person,equals
![image](https://user-images.githubusercontent.com/1567693/104324711-1405c380-54e8-11eb-9bfb-90af2e00ee73.png)

On the corresponding page on dspace demo, that doesn't happen: https://dspace7-demo.atmire.com/search?page=1&query=person&f.entityType=Person,equals

This happens not just in search but everywhere person list elements are rendered

### 4. Item pages show the list element instead
Go to e.g. http://localhost:4000/items/e02f3934-56d6-4aad-ba36-fa5a1b815d4e, and you'll see something like this
![image](https://user-images.githubusercontent.com/1567693/104325615-1ddbf680-54e9-11eb-9b65-551449740675.png)

### 5. The administer workflow grid view shows the list elements instead
On the administer workflow page, the grid view uses list elements. This only happens here, not on any other pages with a grid view as far as I can tell.

## Addressing issues with dynamically loaded components
Issues 3, 4 and 5 are likely related, and related to the issue with the menu items we created a workaround for using switches. However using switches for _all_ dynamically loaded components doesn't seem like a great solution, even if it is only temporary. 

Perhaps the reason for these issues is the removal of the entryComponents arrays. I know the new angular guidelines say we no longer need them. But in order to ensure the decorators on our model classes are loaded in time, we had to [add them to the module](https://github.com/DSpace/dspace-angular/blob/main/src/app/core/core.module.ts#L303-L358). That models array doesn't do anything, it's only there in order to force angular to load those classes, and see the decorators. 

Perhaps entryComponents did something similar for dynamically loaded components, and now that it's gone we need another array that doesn't do anything besides refer to all dynamically loaded components?

atarix83 commented:
@artlowel @tdonohue 
I should have fixed the issue with decorators. As anticipated Art by slack, I've added a static method to modules (where we had entryComponents) that allows to load components with decorators. For example :
```
@NgModule({
  imports: [
    CommonModule,
    SharedModule
  ],
  declarations: [
    ...COMPONENTS,
  ],
  exports: [
    ...COMPONENTS
  ]
})
export class ResearchEntitiesModule {
  static withEntryComponents() {
    return {
      ngModule: ResearchEntitiesModule,
      providers: COMPONENTS.map((component) => ({provide: component}))
    };
  }
}
```
and the import of the module is like
```
ResearchEntitiesModule.withEntryComponents()
```

Could you please give it a try to check if I've addressed all decorators?

artlowel commented:
Thanks @atarix83 that fixed most of these issues.

However certain Item pages still had issue 4 from [my previous comment](https://github.com/DSpace/dspace-angular/pull/970#pullrequestreview-566204233). 
E.g. http://localhost:4000/items/6c33230e-5d6a-4f98-b4e7-3a5c39982897 
Their components had to be added to the module in a similar way. I did so in: https://github.com/atmire/dspace-angular/commit/aa4ec79da08b7c82b25291811829269fab711367

I also reverted the changes I made to the menu components in https://github.com/atmire/dspace-angular/commit/d5559aa9adea6b57a4679a21364e6dc1422e97f3, and added them to the module in a similar way in https://github.com/atmire/dspace-angular/commit/498ba215313dcbd42845e00ea9668e709f414c14

Issue 2 from [my previous comment]([my previous comment](https://github.com/DSpace/dspace-angular/pull/970#pullrequestreview-566204233)) still exists. Since that seems related to the ng-dynamic forms upgrade perhaps you can still take a look? It's likely similar to the changes you made for the submission

I'll take another look at the angulartics injection issue in the mean time

artlowel commented:
I fixed the inject issue by mocking the Angulartics2 Service as well on the server: https://github.com/atmire/dspace-angular/commit/17b3be6757a1cd4ce54727864ddfb2720b999f19

I also noticed the administrative search page still didn't work with dynamic components. That was fixed by importing the modules containing those components: https://github.com/atmire/dspace-angular/commit/b2f830e0d63d3c7c0e454aebd4214d18884e51d7

atarix83 commented:
thanks @artlowel I've picked you fixes and resolve the Issue 2

artlowel reviewed (approved):
Thanks @atarix83!

tdonohue reviewed (approved):
👍 Looks good to me as well.  I've re-tested today and I'm **not** seeing any obvious issues in the app while running this PR.  All pages seem to load and work as expected.  I've also done some basic edits & tested the submission UI and they all seem to be running fine as well.

We do need a ticket created to address the temporarily disabled tests (@atarix83 will create it).  Beyond that, this looks ready to merge to me.  If any other issues arise, we can address them in future PRs as needed.

tdonohue commented:
Merging as this is at +2.  The only failure here is related to the (temporarily) disabled tests -- @atarix83 will create a ticket to ensure those don't get forgotten.
